gennet-module/gennet.py

```diff
@@ -256,33 +296,68 @@ def dict_to_arrays(dic):
     return keys, vals
 
 
+# Check for range failures in a list
+def range_fails(lst, min=0, max=100):
+    return any(x < min or x > max for x in lst)
+
+
+# Check for sum failures in a list
+def sum_fails(lst, sum_expected=100):
+    return not sum(lst) == sum_expected
+
+
+# Construct the nodeType from the trait
+def traits_to_nodeType(s):
+    return nodeType(s.split(':')[0])
+
+# Validate the traits distribution (stick to percentages: num nodes may vary post generation)
+def validate_traits_distribution(traits_distribution):
+    traits, traits_freq = dict_to_arrays(traits_distribution)
+    if range_fails(traits_freq, max=100) :
+        raise ValueError(f"{traits_distribution} : invalid percentage (>{100} or <0)")
+    if sum_fails(traits_freq, sum_expected=100) :
+        raise ValueError(
+                f"{traits_distribution} : percentages do not sum to {100}")
+    for s in traits :
+        traits_list = s.split(":")
+        for t in traits_list :
+            if t not in Trait and not os.path.exists(f"{TRAITS_DIR}/{t}.toml") :
+                raise ValueError(f"{traits_distribution} : unknown trait {t} in {s}")
+
+
+# Extract the traits distribution
+def extract_traits_distribution(traits_distribution):
+    return dict_to_arrays(traits_distribution)
+
+
 # Generate a list of nodeType enums that respects the node type distribution
-def generate_node_types(node_type_distribution, G):
+def generate_traits_distribution(node_type_distribution, G):
     num_nodes = G.number_of_nodes()
-    nodes, node_probs = dict_to_arrays(node_type_distribution)
-    node_types_str = random.choices(nodes, weights=node_probs, k=num_nodes)
-    node_types_enum = [nodeType(s) for s in node_types_str]
-    return node_types_enum
-
-# Inverts a dictionary of lists
-def invert_dict_of_list(d):
-    inv = {}
+    nodes, node_percentage = dict_to_arrays(node_type_distribution)
+    traits_distribution = []
+    for i, n in enumerate(nodes) :
+       traits_distribution +=  [nodes[i]] * math.ceil(node_percentage[i] * num_nodes/100)
+    random.shuffle(traits_distribution)
+    return traits_distribution
+
+
+# Inverts a dictionary of lists (of lists/tuples) 
+def invert_dict_of_list(d, idx=0):
+    inv = defaultdict(list)
     for key, val in d.items():
-        if val not in inv:
-            inv[val] = [key]
-        else:
-            inv[val].append(key)
+        inv[val[idx]].append(key)
     return inv
```

Comment:
Is the `idx` really needed?
I assume each node will be part of a subnetwork. 
If for some reason in the future this can be changed, I don't really how the idx would tackle that. Just openning a discussion about this.

Reply:
because this function is used to reverse node2container as well: node2container's values are tuples (port shift, container id). this generalised inversion would work for any n-tuple/list, as mentiond  in the comments.